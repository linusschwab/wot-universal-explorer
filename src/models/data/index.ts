export {Data} from './Data'
export {InputData} from './InputData'
export {OutputData} from './OutputData'
