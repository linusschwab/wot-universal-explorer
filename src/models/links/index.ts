export {Link} from './Link';
export {HTTPLink} from './HTTPLink';
export {COAPLink} from './COAPLink';
export {Operation} from './Operation';
