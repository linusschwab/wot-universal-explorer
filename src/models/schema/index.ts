export {DataSchema} from './DataSchema'
export {InputSchema} from './InputSchema'
export {OutputSchema} from './OutputSchema'
