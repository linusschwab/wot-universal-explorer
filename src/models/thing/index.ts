export {Thing} from './Thing';
export {ThingsManager} from './ThingsManager';
