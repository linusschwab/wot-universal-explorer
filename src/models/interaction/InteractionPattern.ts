import {Link} from "../links";


export class InteractionPattern {

    public name: string;
    public base: string;
    public links: Link[];

    constructor(name: string) {
        this.name = name;
        this.links = [];
    }

    public registerLink(link: Link) {
        this.links.push(link);
        link.interaction = this;
    }

    public toString() {
        return this.name + ' interaction';
    }

}
