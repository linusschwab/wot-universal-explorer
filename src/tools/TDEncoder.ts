import {Thing} from "../models/thing";
import {Action, InteractionPattern, Property, Event} from "../models/interactions";
import {Link} from "../models/links";
import {DataSchema} from "../models/schema";
import {App} from "../App";


export class TDEncoder {

    public static encode(thing: Thing) {
        return JSON.stringify(thing, (key: string, value: Object) => {
            if (value instanceof Thing) {
                return this.thing(value);
            }
            if (value instanceof InteractionPattern) {
                return this.interaction(value);
            }

            return value;
        }, 4);
    }

    private static thing(thing: Thing) {
        return {
            "@context": ["https://w3c.github.io/wot/w3c-wot-td-context.jsonld",
                "https://w3c.github.io/wot/w3c-wot-common-context.jsonld"],
            "@type": [thing.type],
            "name": thing.name,
            "base": App.url + '/things',
            "interaction": thing.interactions
        };
    }

    // Interactions
    private static interaction(interaction: InteractionPattern) {
        if (interaction instanceof Property) {
            return this.property(interaction);
        }
        else if (interaction instanceof Action) {
            return this.action(interaction);
        }
        else if (interaction instanceof Event) {
            return this.event(interaction);
        }

        // TODO: Throw error instead?
        return null;
    }

    private static property(property: Property) {
        return {
            "@type": ["Property"],
            "name": property.name,
            "schema": this.schema(property.schema),
            "writable": property.writable,
            "observable": property.observable,
            "form": this.form(property.url)
        }
    }

    private static action(action: Action) {
        return {
            "@type": ["Action"],
            "name": action.name,
            "inputSchema": this.schema(action.inputSchema),
            "form": this.form(action.url)
        }
    }

    private static event(event: Event) {
        return {
            "@type": ["Event"],
            "name": event.name,
            "schema": this.schema(event.schema),
            "form": this.form(event.url)
        }
    }

    private static schema(schema: DataSchema) {
        if (schema) {
            return {
                "type": schema.type
            }
        } else {
            return null;
        }
    }

    private static form(url: string) {
        return [
            {
                "href": url,
                "mediaType": "application/json"
            }
        ]
    }
}
