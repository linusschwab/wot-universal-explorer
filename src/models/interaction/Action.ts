import {InteractionPattern} from "./InteractionPattern";
import {InputSchema, OutputSchema} from "../schema";


export class Action extends InteractionPattern {

    public outputSchema: OutputSchema;
    public inputSchema: InputSchema;

    constructor(name: string, outputSchema: OutputSchema, inputSchema: InputSchema) {
        super(name);
        this.outputSchema = outputSchema;
        this.inputSchema = inputSchema;
    }

    public async invoke() {
        // TODO: Choose correct link
        return this.links[0].execute();
    }

    public toString() {
        return this.name + ' action';
    }

}
