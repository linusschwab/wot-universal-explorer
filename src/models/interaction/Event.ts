import {InteractionPattern} from "./InteractionPattern";
import {DataSchema} from "../schema";
import {Operation} from "../links";


export class Event extends InteractionPattern {

    public schema: DataSchema;

    constructor(name: string, schema: DataSchema) {
        super(name);
        this.schema = schema;
    }

    public toString() {
        return 'Event ' + this.name;
    }

    public get url() {
        return '/' + this.thing.slug + '/events/' + this.slug;
    }

    public get operations(): Operation[] {
        return [new Operation('post', 'Subscribe to ' + this.toString(), this)];
    }
}
