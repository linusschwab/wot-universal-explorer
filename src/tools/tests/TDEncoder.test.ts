import {Thing} from "../../models/thing";
import {Action, Property} from "../../models/interaction";
import {DataSchema, InputSchema, OutputSchema} from "../../models/schema";
import {HTTPLink, Link} from "../../models/links";
import {TDEncoder} from "../TDEncoder";


test('thing with base, 1 writable property and 1 action', () => {
    // Prepare test thing
    const thing = new Thing('Testthing', 'Thing', 'http://localhost');

    const property = new Property('testProperty', new DataSchema('integer', true), true, false);
    property.registerLink(new HTTPLink('/testproperty', thing.base, 'application/json'));
    thing.registerInteraction(property);

    const action = new Action('testAction', new InputSchema('boolean'), new OutputSchema('boolean'));
    action.registerLink(new HTTPLink('/testaction', thing.base, 'application/json'));
    thing.registerInteraction(action);

    // Encode
    const td = TDEncoder.encode(thing);
    expect(td).toMatchSnapshot();
});

test('thing without base, 1 property with multiple links', () => {
    // Prepare test thing
    const thing = new Thing('Testthing', 'Thing');

    const property = new Property('testProperty', new DataSchema('integer', true), true, false);
    property.registerLink(new HTTPLink('http://link1.com/testproperty', '', 'application/json'));
    property.registerLink(new HTTPLink('http://link2.com/testproperty', '', 'application/json'));
    thing.registerInteraction(property);

    // Encode
    const td = TDEncoder.encode(thing);
    expect(td).toMatchSnapshot();
});
