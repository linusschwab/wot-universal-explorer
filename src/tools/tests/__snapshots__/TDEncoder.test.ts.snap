// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thing with base, 1 writable property and 1 action 1`] = `
"{
    \\"@context\\": [
        \\"https://w3c.github.io/wot/w3c-wot-td-context.jsonld\\",
        \\"https://w3c.github.io/wot/w3c-wot-common-context.jsonld\\"
    ],
    \\"@type\\": [
        \\"Thing\\"
    ],
    \\"name\\": \\"Testthing\\",
    \\"base\\": \\"http://localhost:5000/things\\",
    \\"interaction\\": [
        {
            \\"@type\\": [
                \\"Property\\"
            ],
            \\"name\\": \\"testProperty\\",
            \\"schema\\": {
                \\"type\\": \\"integer\\"
            },
            \\"writable\\": true,
            \\"observable\\": false,
            \\"form\\": [
                {
                    \\"href\\": \\"/testthing/properties/testproperty\\",
                    \\"mediaType\\": \\"application/json\\"
                }
            ]
        },
        {
            \\"@type\\": [
                \\"Action\\"
            ],
            \\"name\\": \\"testAction\\",
            \\"inputSchema\\": {
                \\"type\\": \\"boolean\\"
            },
            \\"form\\": [
                {
                    \\"href\\": \\"/testthing/actions/testaction\\",
                    \\"mediaType\\": \\"application/json\\"
                }
            ]
        }
    ]
}"
`;

exports[`thing without base, 1 property with multiple links 1`] = `
"{
    \\"@context\\": [
        \\"https://w3c.github.io/wot/w3c-wot-td-context.jsonld\\",
        \\"https://w3c.github.io/wot/w3c-wot-common-context.jsonld\\"
    ],
    \\"@type\\": [
        \\"Thing\\"
    ],
    \\"name\\": \\"Testthing\\",
    \\"base\\": \\"http://localhost:5000/things\\",
    \\"interaction\\": [
        {
            \\"@type\\": [
                \\"Property\\"
            ],
            \\"name\\": \\"testProperty\\",
            \\"schema\\": {
                \\"type\\": \\"integer\\"
            },
            \\"writable\\": true,
            \\"observable\\": false,
            \\"form\\": [
                {
                    \\"href\\": \\"/testthing/properties/testproperty\\",
                    \\"mediaType\\": \\"application/json\\"
                }
            ]
        }
    ]
}"
`;
